src/linker/Linker/Tracer.cs

```diff
@@ -28,72 +28,62 @@
 
 using System.Collections.Generic;
 using System.IO;
-using System.IO.Compression;
-using System.Xml;
-using Mono.Cecil;
-using Mono.Linker.Steps;
 
 namespace Mono.Linker
 {
-	public class Tracer {
+	public class Tracer
+	{
 		public const string DefaultDependenciesFileName = "linker-dependencies.xml.gz";
 
 		public string DependenciesFileName { get; set; } = DefaultDependenciesFileName;
 
 		protected readonly LinkContext context;
 
-		Stack<object> dependency_stack;
-		System.Xml.XmlWriter writer;
-		Stream stream;
+		private XmlDependencyRecorder xmlDependencyRecorder;
 
-		public Tracer (LinkContext context) => this.context = context;
+		Stack<object> dependency_stack;
+		List<IDependencyRecorder> recorders;
 
-		public void Start ()
+		public Tracer (LinkContext context)
 		{
+			this.context = context;
 			dependency_stack = new Stack<object> ();
-			System.Xml.XmlWriterSettings settings = new System.Xml.XmlWriterSettings {
-				Indent = true,
-				IndentChars = "\t"
-			};
+		}
 
+		public void Start ()
+		{
 			if (string.IsNullOrEmpty (Path.GetDirectoryName (DependenciesFileName)) && !string.IsNullOrEmpty (context.OutputDirectory)) {
 				DependenciesFileName = Path.Combine (context.OutputDirectory, DependenciesFileName);
 				Directory.CreateDirectory (context.OutputDirectory);
 			}
 
-			var depsFile = File.OpenWrite (DependenciesFileName);
+			xmlDependencyRecorder = new XmlDependencyRecorder (DependenciesFileName, context);
```

Comment:
I think xmlDependencyRecorder should not be part of this type. It should be set externally